pallets/committee/src/lib.rs

```diff
@@ -351,26 +362,35 @@ pub mod pallet {
 		/// This can only be called by members of the committee.
 		/// Successfully cast votes will be recorded in the state and a proposal
 		/// meeting voting requirements can be executed.
+		#[transactional]
 		#[pallet::weight((T::WeightInfo::vote(), DispatchClass::Operational))]
 		pub fn vote(origin: OriginFor<T>, proposal_hash: HashFor<T>, vote: VoteKind) -> DispatchResult {
-			// Only members can vote
 			let voter = Self::ensure_member(origin)?;
 
-			Votes::<T>::try_mutate(&proposal_hash, |votes| {
-				if let Some(votes) = votes {
-					// Can only vote within the allowed range of blocks for this proposal
-					ensure!(Self::within_voting_period(votes), Error::<T>::NotInVotingPeriod);
-					// members can vote only once
-					ensure!(!votes.has_voted(&voter.account_id), Error::<T>::DuplicateVote);
-					votes.cast_vote(MemberVote::new(voter.clone(), vote.clone())); // mutates votes in place
-					Self::deposit_event(Event::VoteCast(voter, proposal_hash, vote));
-					Ok(())
-				} else {
-					Err(Error::<T>::NoProposalWithHash)
+			VotingEligibility::<T>::try_mutate(&voter.account_id, |maybe_eligibility| -> DispatchResult {
+				if maybe_eligibility
+					.take()
+					.filter(|block_number| frame_system::Pallet::<T>::block_number() < *block_number)
```

Comment:
should we check for `0` here?

Reply:
here is a `take` before `filter`, `if frame_system::Pallet::<T>::block_number() >= block_number {}`, the storage will be removed automatically

Reply:
the `VotingEligibility::<T>::try_mutate` is unnecessary anyway, isn't it?
we should `get` it instead and then do the check.

Reply:
ur right, used to try using `try_mutate` to avoid unexpected storage modification, since there is a `#[transactional]` tag on this function already, we can take and check directly

Reply:
We don't want to remove the eligibility timestamp. we should just check it instead. so no modifications of the eligibility map.